src/chain.js

```diff
@@ -0,0 +1,88 @@
+const http = require('http');
+
+/**
+ * @param {http.ClientRequest} request
+ * @param {net.Socket} source
+ * @param {buffer.Buffer} head
+ * @param {*} handlerOpts
+ * @param {*} server
+ */
+const chain = (request, source, head, handlerOpts, server) => {
+    if (head.length > 0) {
+        throw new Error(`Unexpected data on CONNECT: ${head.length} bytes`);
+    }
+
+    const { upstreamProxyUrlParsed: proxy } = handlerOpts;
+
+    const options = {
+        method: 'CONNECT',
+        hostname: proxy.hostname,
+        port: proxy.port,
+        path: request.url,
+        headers: [
+            'host',
+            request.url,
+        ],
+    };
+
+    if (proxy.username || proxy.password) {
+        const auth = `${proxy.username}:${proxy.password}`;
```

Comment:
What is either `username` or `password` is null? Then the auth string will contain "null".

Reply:
In WHATWG URL it's never null, it's an empty string.

Reply:
With TypeScript we won't have to worry `can this be null?` :)